### PR TITLE
Resolver: don't use one global instance and implement caching

### PR DESCRIPTION
**Pull request by humitos**
The `resolver.resolve` method that resolves the URL **always** check for the canonical `Domain` of the canonical `Project`. This means that resolving the URLs for multiple `Version` objects results in querying the exact query over and over again.

Instead of repeating this query, we are passing the `Domain` object to `resolver.resolve` method to avoid hiting the DB again since we already known what's the canonical domain for that project, in case it has one.

I found this issue while tracking down why there were around ~150 queries to the `projects_domain` table when hitting `/_/addons/` API endpoint on New Relic.

### CODE REVIEW

**Comment by humitos** [comment]
I modified my local settings as follows to get all the DB queries performed per request:

```diff
diff --git a/readthedocs/settings/docker_compose.py b/readthedocs/settings/docker_compose.py
index abf884244..f92d111f0 100644
--- a/readthedocs/settings/docker_compose.py
+++ b/readthedocs/settings/docker_compose.py
@@ -134,6 +134,11 @@ class DockerBaseSettings(CommunityBaseSettings):
                     "handlers": ["null"],
                     "propagate": False,
                 },
+                'django.db.backends': {
+                    'handlers': ['console'],
+                    'level': 'DEBUG',
+                    'propagate': True,
+                },
             }
         )
         return logging
```

**Comment by ericholscher** [comment]
I think this should probably be on top of https://github.com/readthedocs/readthedocs.org/pull/10813 which is a pretty big refactor of this code that's already outstanding. I think it does some of this work already, but I haven't had time to 100% figure out what all is going on with it.

**Comment by stsewd** [comment]
If you are resolving versions from the same project, you can just do something similar to

https://github.com/readthedocs/readthedocs.org/blob/f75e4de3c58b2421654f529ada9556ae4147fffa/readthedocs/projects/views/mixins.py#L82-L95

- First resolve the domain, since all versions share the same main project, they all share the same domain.
- Resolve the path for each version and use the resolved domain from before.

In https://github.com/readthedocs/readthedocs.org/pull/10813 we can expose the `_get_project_domain` method so we don't need to resolve the path for the first call to `resolve(project)`.

**Comment by humitos** [comment]
@stsewd 

> First resolve the domain, since all versions share the same main project, they all share the same domain.
> Resolve the path for each version and use the resolved domain from before.

I'd prefer to not manipulate the URLs by hand and strictly rely on the `resolver.resolve` method. Mainly, because that's the source of truth and the method we will be adding logic changes and optimizations.

I don't want to have a customized way of resolving the URLs that I have to maintain separately from the resolver. The resolver should be the only one that resolves these URLs and the one we all benefit from each time we slightly optimize, IMO.

**Comment by humitos** [comment]
@stsewd 

> In #10813 we can expose the `_get_project_domain` method so we don't need to resolve the path for the first call to `resolve(project)`.

Yes. Being able to call `resolver._get_project_domain` without being a private method would be good 👍🏼

**Comment by humitos** [comment]
@ericholscher 

> I think this should probably be on top of #10813 which is a pretty big refactor of this code that's already outstanding

I can continue working on this PR, once #10813 gets merged.

**Comment by humitos** [comment]
I updated this PR with a completely different approach: using `functools.lru_cache` on `resolver._get_project_domain`. This caches the domain resolution in memory and it re-uses it to resolve all the following URLs. So, we went from "1 db query per version" to "1 db query no matter how many  versions the project has" 💯

**Comment by humitos** [comment]
> I'm happy to just add a `cache_domain=False` keyword attribute to `resolver.resolve` to cache the resolution of the project's domain and pass it only where it's needed (e.g. Addons API response)

I've opened a PR with this proposal at https://github.com/readthedocs/readthedocs.org/pull/10878. It looks way simpler and less intrusive with the rest of the code.

**Review by ericholscher — APPROVED** [review]
I think this approach looks cleaner in the code, and agree with @stsewd (https://github.com/readthedocs/readthedocs.org/pull/10878#issuecomment-1785638238) that using the Python cache and no argument is much better. I dislike adding complexity to the codebase just for testing, so if the tests here aren't way worse, I think this approach is better.

**Comment by humitos** [comment]
I followed @stsewd's suggestions from https://github.com/readthedocs/readthedocs.org/pull/10878#issuecomment-1786182309 and updated the code to use a different instance per method/function/module/etc depending on how much cache we want to share.

Now, we are instantiating a new instance each time we need to use the resolver. However, if the caller knows there is going to be data that could be cached/shared inside the function (because it uses a loop or similar), the caller can re-use the same instance to avoid doing multiple calls to the DB for the same values.

I feel more comfortable with this approach. It looks simpler and easier to understand. Thanks for the suggestions here 👍🏼

File: readthedocs/core/resolver.py [hunk 1]
```diff
@@ -352,8 +368,5 @@ def _use_cname(self, project):
         """Test if to allow direct serving for project on CNAME."""
         return bool(get_feature(project, feature_type=TYPE_CNAME))
 
-
-resolver = Resolver()
-resolve_path = resolver.resolve_path
-resolve_domain = resolver.get_domain_without_protocol
-resolve = resolver.resolve
+    def clean_cache(self):
+        self._get_project_domain.cache_clear()
```
Comment (stsewd) [inline readthedocs/core/resolver.py:372]: Don't think we need this anymore

**Review by stsewd — APPROVED** [review]

**Comment by humitos** [comment]
> This makes me wonder more about trying to share this cache in memory across all calls to the resolver. Like what is we used a global instance, and kept an LRU cache of ~1000 calls in memory...

I thought about this, but since the LRU cache doesn't have a timeout, it could be a problem when there is an update on the objects (e.g. `Domain`) that modify the resulting value of the cached method. In that case, _each web instance_ will have to execute the cached method 1000 times with different values to get the new result for the object updated. It could be even pretty confusing if one web instance has already updated the cached value with the new result and another instance has not. This will result in the same project returning different values in different requests: one response could return the custom domain being the canonical domain and the immediate next response could return `.readthedocs.io`.

**Pull request merged**